great_expectations/experimental/datasources/data_asset/data_connector/data_connector.py

```diff
@@ -0,0 +1,190 @@
+from __future__ import annotations
+
+import logging
+import pathlib
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+from great_expectations.core.id_dict import BatchSpec
+
+if TYPE_CHECKING:
+    from great_expectations.core.batch import BatchDefinition
+    from great_expectations.experimental.datasources.interfaces import BatchRequest
+
+
+logger = logging.getLogger(__name__)
+
+
+# noinspection SpellCheckingInspection
+class DataConnector:
+    """The base class for all Data Connectors.
+
+    Data Connectors produce identifying information, called Batch Specs, that Execution Engines
+    can use to get individual batches of data. They add flexibility in how to obtain data
+    such as with time-based partitioning, downsampling, or other techniques appropriate
+    for the Datasource.
+
+    For example, a DataConnector could produce a SQL query that logically represents "rows in
+    the Events table with a timestamp on February 7, 2012," which an SqlAlchemy Datasource
+    could use to materialize a SqlAlchemy Dataset corresponding to that Batch of data and
+    ready for validation.
+
+    A Batch is a sample from a data asset, sliced according to a particular rule. For example,
+    an hourly slide of the Events table or “most recent Users records.” It is the primary
+    unit of validation in the Great Expectations Data Context. Batches include metadata that
+    identifies how they were constructed--the same Batch Spec assembled by the data connector.
+    While not every Datasource will enable re-fetching a specific batch of data, GX can store
+    snapshots of batches or store metadata from an external data version control system.
+
+    Args:
+        name: The name of the DataConnector instance
+        datasource_name: The name of the Datasource associated with this DataConnector instance
+        data_asset_name: The name of the DataAsset using this DataConnector instance
+    """
+
+    def __init__(
+        self,
+        name: str,
+        datasource_name: str,
```

Comment:
We need to stop referencing names and start referencing the actual types. I think this could dramatically simplify much of the GX codebase

Reply:
@donaldheppner Actually, in `DataConnector` passing the entities that it needs is cleaner then passing the entire `DataAsset` object to it, because the implication of the latter approach would be that the `DataConnector` needs to call methods on `DataAsset`, which not the case.  Please note that `datasource_name` is essential for a part of `DataConnector` functionality that matches/filters different `BatchRequest` objects, based on its key properties.  Ported from V3, this is a more formal implementation for what is currently in `_fully_specified_batch_requests_with_path()`, generalizing the language of "fully-specified `BatchRequest`" and "partially-specified `BatchRequest`" to querying `BatchDefinitions` using a rich language of not only the keys, but many other criteria, including full `Python` indexing and slicing as well as being able to supply custom function (`lambda`).

Reply:
If an instance of a data connector is specific to a datasource, is this still the case? If not, why not?